pkg/cloudlogging/puller_test.go

```diff
@@ -98,9 +100,110 @@ func TestPull(t *testing.T) {
 				t.Errorf("Process(%+v) got result diff (-want, +got): %v", tc.name, diff)
 			}
 
-			if diff := cmp.Diff(tc.wantReq, tc.server.req, protocmp.Transform()); diff != "" {
+			if diff := cmp.Diff(tc.wantReq, tc.server.listReq, protocmp.Transform()); diff != "" {
+				t.Errorf("Process(%+v) got request diff (-want, +got): %v", tc.name, diff)
+			}
+		})
+	}
+}
+
+func TestStreamPull(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name          string
+		server        *fakeServer
+		filter        string
+		wantReq       *loggingpb.TailLogEntriesRequest
+		wantResult    []*loggingpb.LogEntry
+		wantNum       int
+		wantErrSubstr string
+	}{
+		{
+			name:       "success",
+			filter:     "test-filter",
+			server:     &fakeServer{},
+			wantResult: []*loggingpb.LogEntry{{LogName: "test-0"}, {LogName: "test-1"}},
+			wantReq: &loggingpb.TailLogEntriesRequest{
+				ResourceNames: []string{testResource},
+				Filter:        "test-filter",
+			},
+			wantNum: 2,
+		},
+		{
+			name:   "failed_to_pull",
+			filter: "test-filter",
+			server: &fakeServer{
+				injectedErr: fmt.Errorf("injected error"),
+			},
+			wantReq: &loggingpb.TailLogEntriesRequest{
+				ResourceNames: []string{testResource},
+				Filter:        "test-filter",
+			},
+			wantErrSubstr: "injected error",
+			wantNum:       0,
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			ch := make(chan *loggingpb.LogEntry)
+			var gotLogs []*loggingpb.LogEntry
+			done := make(chan struct{}, 1)
+			t.Cleanup(func() {
+				close(done)
+			})
+
+			ctx, cancel := context.WithCancel(context.Background())
+
+			go func() {
+				for l := range ch {
+					gotLogs = append(gotLogs, l)
+					if len(gotLogs) == tc.wantNum {
+						cancel() // got enough logs, we can stop now
+						break
+					}
+				}
+			}()
+
+			// set up fake client
+			fakeClient := setupFakeClient(t, ctx, tc.server)
+			p := NewPuller(
+				ctx,
+				fakeClient,
+				testResource,
+				WithRetry(retry.WithMaxRetries(0, retry.NewFibonacci(500*time.Millisecond))),
+			)
+			var gotPullErr error
+			var gotCloseClientErr error
+
+			go func() {
+				defer close(ch)
+				gotPullErr, gotCloseClientErr = p.StreamPull(ctx, tc.filter, ch)
+				cancel()
```

Comment:
I'm not sure if this cancel is required. Can you explain?

Reply:
Removed.

Reply:
I see. Can you comment?

Reply:
Done.